## 구현,실험 전체적인 설명 및 분석 

In [1]:
!pip install torch==1.8.1
!pip install torchtext==0.9.1

     |████████████████████████████████| 804.1 MB 1.9 kB/s 
  Attempting uninstall: torch
    Found existing installation: torch 1.9.1
    Uninstalling torch-1.9.1:
      Successfully uninstalled torch-1.9.1
     |████████████████████████████████| 7.1 MB 889 kB/s 
  Attempting uninstall: torchtext
    Found existing installation: torchtext 0.10.1
    Uninstalling torchtext-0.10.1:
      Successfully uninstalled torchtext-0.10.1


In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib as mpl
import matplotlib.ticker as ticker
from numpy import array
from nltk.tree import Tree
import re

import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
import torchtext
from torchtext.legacy import data
from torchtext.legacy.data import Field
from torchtext.legacy.data import TabularDataset, BucketIterator

import random
import math
import time

import warnings
warnings.filterwarnings(action='ignore')

In [3]:
# file processing
en_file = open('../input/koenen/ko-en.en.parse.syn.txt','r', encoding='UTF8')
english = en_file.readlines() 

ko_File = open('../input/koenko/ko-en.ko.parse.txt','r', encoding='UTF8')
korean = ko_File.readlines()

In [4]:
# function to read raw text file
def read_text(filename): 
    # open the file 
    file = open('../input/koenko/ko-en.ko.parse.txt', mode='rt', encoding='utf-8') 
        
    # read all text 
    text = file.read() 
    file.close() 
    return text
    
# split text into sentences # 이 노트북에서는 형태소로 나뉜 자료로 사용했습니다.
def to_lines(text): 
    sents = text.strip().split('\n\n')  # 한 문장 내 분리된 채 사용할 때
    return sents

def to_other_lines(text): 
    sents = text.strip().split('\n\n') 
    sents = [i.split('\t') for i in sents]  #이거 입력하면 어절별로 처리 가능
    return sents

In [5]:
# 어절별
#ko_eng_other = to_other_lines(data) 
#ko_eng_other = pd.DataFrame(ko_eng_other)
#ko_eng_other.head()
#ko_eng_other.iloc[:,::3] # 필요한 문장이 담긴 컬럼만 추출

In [6]:
# 데이터프레임으로 만들기(한국어 문장)
data = read_text("ko-en.ko.parse.txt") 
ko_eng_lines = to_lines(data) 
ko_eng = pd.DataFrame(ko_eng_lines, columns = ["read_lines"])

pd.set_option('display.max_colwidth', 250)
ko_eng.head()

read_lines
0  <id 1>\n<sent 1>\n1\t2\tNP\t777/SN\n2\t6\tNP_SBJ\t항공편/NNG|은/JX\n3\t4\tNP\t1/SN|시간/NNG\n4\t6\tNP_AJT\t동안/NNG\n5\t6\tNP_AJT\t지상/NNG|에/JKB\n6\t7\tVP\t머물/VV|게/EC\n7\t0\tVP\t되/VV|ㅂ니다/EF|./SF\n</sent>\n</id>
1                   <id 2>\n<sent 1>\n1\t2\tNP\t777/SN\n2\t6\tNP_SBJ\t항공편/NNG|은/JX\n3\t4\tNP\t3/SN|시간/NNG\n4\t6\tNP_AJT\t동안/NNG\n5\t6\tNP_AJT\t지상/NNG|에/JKB\n6\t0\tVP\t있/VV|겠/EP|습니다/EF|./SF\n</sent>\n</id>
2                                                                <id 3>\n<sent 1>\n1\t2\tNP\t1/SN|,/SP|000/SN|달러/NNB\n2\t3\tNP_SBJ\t여행자/NNG|수표/NNG|가/JKS\n3\t0\tVP\t필요/NNG|하/XSA|ㅂ니다/EF|./SF\n</sent>\n</id>
3                                                                        <id 4>\n<sent 1>\n1\t3\tNP_SBJ\t1/SN|,/SP|250/SN|원/NNB|이/JKS\n2\t3\tNP\t공식/NNG\n3\t0\tVNP\t환율/NNG|이/VCP|ㅂ니다/EF|./SF\n</sent>\n</id>
4             <id 5>\n<sent 1>\n1\t2\tNP\t100/SN|달러/NNB\n2\t3\tNP_CNJ\t3/SN|장/NNB|과/JC\n3\t5\tNP_SBJ\t나머지/NNG|는/JX\n4\t5\tNP_AJT\t20/SN|달러/NNB|권/XSN|으로/JKB\n5\t0\tVP\t주/VV|시/EP|ㅂ시오/EF|./SF\n</sent>\n</id>

In [7]:
# 데이터프레임으로 만들기(영어 문장)
sentences = []
for a in english:
    first_list = []
    seconde_list = []
    t = Tree.fromstring(a)
    t.pos()
    for i in t.pos():
        a,b = i
        a += ' '
        b += ' '
        first_list.append(a)
        seconde_list.append(b)
        row = (first_list, seconde_list)
    sentences.append((''.join(row[0]), ''.join(row[1])))
    
english_df = pd.DataFrame(sentences, columns = ["English_list", "English_POS"])
english_df.head()

English_list  \
0                                 Flight 007 will stay on the ground for one hour .    
1                              Flight 017 will stay on the ground for three hours .    
2                                      I need 1,000 dollars in traveler 's checks .    
3                                  The official exchange rate is around 1,250 Won .    
4  Please give me three hundred dollar bills and twenty dollar bills for the rest .    

                                       English_POS  
0               NNP NNP MD VB IN DT NN IN CD NN .   
1              NNP NNP MD VB IN DT NN IN CD NNS .   
2                  PRP VBP CD NNS IN NN POS NNS .   
3                    DT JJ NN NN VBZ IN NNP NNP .   
4  UH VB PRP CD CD NN NNS CC CD NN NNS IN DT NN .

In [8]:
# 병렬 데이터를 만들기 위한 기본 전처리
def basic_cleaning_ko(text):
    text = re.sub(r'[A-Z0-9_]+\t', '', text) # 탭과 결합된 문자 제거
    text = re.sub(r'<id [0-9]+>', '', text) # 불필요한 글자 제거
    text = re.sub(r'</id>', '', text)
    text = re.sub(r'<sent [0-9]+>', '', text)
    text = re.sub(r'</sent>', '', text)
    text = re.sub(r'\n', ' ', text) # 줄바꿈 공백으로 바꿔주기
    text = re.sub(r'\|', ' ', text) # vertical var도 공백으로 바꿔주기
    text = re.findall('(.+?)/[A-Z]+\s*', text) # 품사기호 앞에 있는 단어만 추출
    return ' '.join(text)

def basic_cleaning_en(text):
    text = re.split('(\d+)', text) # 숫자와 글자 분리
    return ' '.join(text)

In [9]:
ko_eng["clean_Korean"] = ko_eng["read_lines"].apply(basic_cleaning_ko)
english_df["clean_English"] = english_df["English_list"].apply(basic_cleaning_en)

In [10]:
# 한국어-영어 병렬 데이터 만들기
ko_trans =ko_eng["clean_Korean"]
en_trans = english_df["clean_English"]
final_df = pd.concat([ko_trans, en_trans], axis=1)
final_df

clean_Korean  \
0                        777 항공편 은 1 시간 동안 지상 에 머물 게 되 ㅂ니다 .   
1                           777 항공편 은 3 시간 동안 지상 에 있 겠 습니다 .   
2                             1 , 000 달러 여행자 수표 가 필요 하 ㅂ니다 .   
3                                  1 , 250 원 이 공식 환율 이 ㅂ니다 .   
4                    100 달러 3 장 과 나머지 는 20 달러 권 으로 주 시 ㅂ시오 .   
...                                                      ...   
330969    저 의 주소 가 지방 으로 되 어 있 는데 , 서울 에서 여권 을 만들 ㄹ 수 있 나요 ?   
330970                               학교 친구 한 명 이 LA 에 있 어요 .   
330971                                        이것 ㄴ 얼마 이 에요 ?   
330972                                        저것 ㄴ 얼마 이 에요 ?   
330973                                        저것 은 얼마 이 에요 ?   

                                                                            clean_English  
0                                    Flight  007  will stay on the ground for one hour .   
1                                 Flight  017  will stay on the ground for three hours .   
2                                       I need  1 , 000  dollars in traveler 's checks .   
3                                   The official exchange rate is around  1 , 250  Won .   
4       Please give me three hundred dollar bills and twenty dollar bills for the rest .   
...                                                                                   ...  
330969  Can I file my passport application in Seoul even if I do n't live in that area ?   
330970                                           I have a school friend in Los Angeles .   
330971                                                                How much is this ?   
330972                                                                How much is that ?   
330973                                                                How much is that ?   

[330974 rows x 2 columns]

In [11]:
# 토큰화 
def tok_english(text): 
    text = text.lower() # 소문자 변환
    text = text.replace("n't", 'not') # 기본적인 축약형 수정
    text = text.replace("'ve", 'have')
    text = text.replace("'m", 'am')
    text = text.replace("'s", ' ')
    text = text.replace("'ll", 'will')
    text = text.replace("'d", 'would')    
    #text = re.sub(r'[^\w\s]', '', text) # 구두점 제거
    text = text.replace('  ', ' ') # 다중 공백 제거
    text = text.replace('  ', ' ') # 두번
    return ''.join(text).split() # 단어 분리

def tok_korean(text):
   # text = re.sub(r'[^\w\s]', '', text) # 구두점 제거 
    text = text.replace('  ', ' ') # 다중 공백 제거
    return ''.join(text).split() # 단어 분리

In [12]:
# 확인
final_df["clean_Korean"].apply(tok_korean).tail()

330969    [저, 의, 주소, 가, 지방, 으로, 되, 어, 있, 는데, ,, 서울, 에서, 여권, 을, 만들, ㄹ, 수, 있, 나요, ?]
330970                                          [학교, 친구, 한, 명, 이, LA, 에, 있, 어요, .]
330971                                                       [이것, ㄴ, 얼마, 이, 에요, ?]
330972                                                       [저것, ㄴ, 얼마, 이, 에요, ?]
330973                                                       [저것, 은, 얼마, 이, 에요, ?]
Name: clean_Korean, dtype: object

In [13]:
# 확인
final_df["clean_English"].apply(tok_english).tail()

330969    [can, i, file, my, passport, application, in, seoul, even, if, i, do, not, live, in, that, area, ?]
330970                                                      [i, have, a, school, friend, in, los, angeles, .]
330971                                                                               [how, much, is, this, ?]
330972                                                                               [how, much, is, that, ?]
330973                                                                               [how, much, is, that, ?]
Name: clean_English, dtype: object

In [14]:
# devide into train and test set

SEED = 42

from sklearn.model_selection import train_test_split
train_data, test_data = train_test_split(final_df, test_size=0.2, shuffle=True, random_state=SEED)

In [15]:
print(len(train_data))
print(len(test_data))

264779
66195


In [16]:
# save to csv file
train_data.to_csv("train.csv")
test_data.to_csv("test.csv")

In [17]:
train_reading = pd.read_csv("./train.csv")
train_reading.head(1) 

Unnamed: 0                  clean_Korean  \
0       41455    이 전화 로 스위스 에 걸 ㄹ 수 있 습니까 ?   

                                   clean_English  
0  Can I call Switzerland with this telephone ?

In [18]:
SEED = 1234

random.seed(SEED)
np.random.seed(SEED)
torch.manual_seed(SEED)
torch.cuda.manual_seed(SEED)
torch.backends.cudnn.deterministic = True

In [19]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

# Convolutional seq2seq

In [20]:
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
import torchtext
from torchtext.legacy import data
from torchtext.legacy.data import Field
from torchtext.legacy.data import TabularDataset, BucketIterator

In [21]:
SRC = Field(tokenize = tok_korean, 
            init_token = '<sos>', 
            eos_token = '<eos>', 
            lower = True, 
            batch_first = True)

TRG = Field(tokenize = tok_english, 
            init_token = '<sos>', 
            eos_token = '<eos>', 
            lower = True,
            batch_first = True)

Raw_datafields = [("Unnamed:0", None), ("src", SRC), ("trg", TRG)]

train_data = data.TabularDataset(
        path='./train.csv',
        format='csv',
        skip_header=True, 
        fields=Raw_datafields) 
test_data = data.TabularDataset(
        path='./test.csv',
        format='csv',
        skip_header=True,
        fields=Raw_datafields)

In [22]:
SRC.build_vocab(train_data, min_freq = 1)
TRG.build_vocab(train_data, min_freq = 1)

BATCH_SIZE = 256

train_iterator, test_iterator = BucketIterator.splits(
    (train_data, test_data), sort = False, repeat = False,
    sort_within_batch =False,
     batch_size = BATCH_SIZE,
     device = device)

In [23]:
class Encoder(nn.Module):
    def __init__(self, 
                 input_dim, 
                 emb_dim, 
                 hid_dim, 
                 n_layers, 
                 kernel_size, 
                 dropout, 
                 device,
                 max_length = 100):
        super().__init__()
        
        assert kernel_size % 2 == 1, "Kernel size must be odd!"
        
        self.device = device       
        self.scale = torch.sqrt(torch.FloatTensor([0.5])).to(device)        
        self.tok_embedding = nn.Embedding(input_dim, emb_dim)
        self.pos_embedding = nn.Embedding(max_length, emb_dim)        
        self.emb2hid = nn.Linear(emb_dim, hid_dim)
        self.hid2emb = nn.Linear(hid_dim, emb_dim)       
        self.convs = nn.ModuleList([nn.Conv1d(in_channels = hid_dim, 
                                              out_channels = 2 * hid_dim, 
                                              kernel_size = kernel_size, 
                                              padding = (kernel_size - 1) // 2)
                                    for _ in range(n_layers)])
        
        self.dropout = nn.Dropout(dropout)
        
    def forward(self, src):
        
        batch_size = src.shape[0]
        src_len = src.shape[1] 
        pos = torch.arange(0, src_len).unsqueeze(0).repeat(batch_size, 1).to(self.device)        
        tok_embedded = self.tok_embedding(src)
        pos_embedded = self.pos_embedding(pos)        
        embedded = self.dropout(tok_embedded + pos_embedded)        
        conv_input = self.emb2hid(embedded)
        conv_input = conv_input.permute(0, 2, 1) 

        for i, conv in enumerate(self.convs):

            conved = conv(self.dropout(conv_input))
            conved = F.glu(conved, dim = 1)
            conved = (conved + conv_input) * self.scale
            conv_input = conved

        conved = self.hid2emb(conved.permute(0, 2, 1))
        combined = (conved + embedded) * self.scale
        
        return conved, combined
    
class Decoder(nn.Module):
    def __init__(self, 
                 output_dim, 
                 emb_dim, 
                 hid_dim, 
                 n_layers, 
                 kernel_size, 
                 dropout, 
                 trg_pad_idx, 
                 device,
                 max_length = 100):
        super().__init__()
        
        self.kernel_size = kernel_size
        self.trg_pad_idx = trg_pad_idx
        self.device = device        
        self.scale = torch.sqrt(torch.FloatTensor([0.5])).to(device)        
        self.tok_embedding = nn.Embedding(output_dim, emb_dim)
        self.pos_embedding = nn.Embedding(max_length, emb_dim)        
        self.emb2hid = nn.Linear(emb_dim, hid_dim)
        self.hid2emb = nn.Linear(hid_dim, emb_dim)        
        self.attn_hid2emb = nn.Linear(hid_dim, emb_dim)
        self.attn_emb2hid = nn.Linear(emb_dim, hid_dim)        
        self.fc_out = nn.Linear(emb_dim, output_dim)        
        self.convs = nn.ModuleList([nn.Conv1d(in_channels = hid_dim, 
                                              out_channels = 2 * hid_dim, 
                                              kernel_size = kernel_size)
                                    for _ in range(n_layers)])
        
        self.dropout = nn.Dropout(dropout)
      
    def calculate_attention(self, embedded, conved, encoder_conved, encoder_combined):

        conved_emb = self.attn_hid2emb(conved.permute(0, 2, 1))        
        combined = (conved_emb + embedded) * self.scale                
        energy = torch.matmul(combined, encoder_conved.permute(0, 2, 1))        
        attention = F.softmax(energy, dim=2)            
        attended_encoding = torch.matmul(attention, encoder_combined)
        attended_encoding = self.attn_emb2hid(attended_encoding)
        attended_combined = (conved + attended_encoding.permute(0, 2, 1)) * self.scale
        
        return attention, attended_combined
        
    def forward(self, trg, encoder_conved, encoder_combined):

        batch_size = trg.shape[0]
        trg_len = trg.shape[1]
        pos = torch.arange(0, trg_len).unsqueeze(0).repeat(batch_size, 1).to(self.device)
        tok_embedded = self.tok_embedding(trg)
        pos_embedded = self.pos_embedding(pos)
        embedded = self.dropout(tok_embedded + pos_embedded)
        conv_input = self.emb2hid(embedded)
        conv_input = conv_input.permute(0, 2, 1)         
        batch_size = conv_input.shape[0]
        hid_dim = conv_input.shape[1]
        
        for i, conv in enumerate(self.convs):

            conv_input = self.dropout(conv_input)
            padding = torch.zeros(batch_size, 
                                  hid_dim, 
                                  self.kernel_size - 1).fill_(self.trg_pad_idx).to(self.device)
                
            padded_conv_input = torch.cat((padding, conv_input), dim = 2) 
            conved = conv(padded_conv_input)
            conved = F.glu(conved, dim = 1)
            attention, conved = self.calculate_attention(embedded, 
                                                         conved, 
                                                         encoder_conved, 
                                                         encoder_combined)

            conved = (conved + conv_input) * self.scale
            conv_input = conved
            
        conved = self.hid2emb(conved.permute(0, 2, 1))            
        output = self.fc_out(self.dropout(conved))
            
        return output, attention
    
class Seq2Seq(nn.Module):
    def __init__(self, encoder, decoder):
        super().__init__()
        
        self.encoder = encoder
        self.decoder = decoder
        
    def forward(self, src, trg):

        encoder_conved, encoder_combined = self.encoder(src)
        output, attention = self.decoder(trg, encoder_conved, encoder_combined)
               
        return output, attention

In [24]:
INPUT_DIM = len(SRC.vocab)
OUTPUT_DIM = len(TRG.vocab)
EMB_DIM = 256
HID_DIM = 512 # each conv. layer has 2 * hid_dim filters
ENC_LAYERS = 10 # number of conv. blocks in encoder
DEC_LAYERS = 10 # number of conv. blocks in decoder
ENC_KERNEL_SIZE = 3 # must be odd!
DEC_KERNEL_SIZE = 3 # can be even or odd
ENC_DROPOUT = 0.25
DEC_DROPOUT = 0.25
TRG_PAD_IDX = TRG.vocab.stoi[TRG.pad_token]
    
enc = Encoder(INPUT_DIM, EMB_DIM, HID_DIM, ENC_LAYERS, ENC_KERNEL_SIZE, ENC_DROPOUT, device)
dec = Decoder(OUTPUT_DIM, EMB_DIM, HID_DIM, DEC_LAYERS, DEC_KERNEL_SIZE, DEC_DROPOUT, TRG_PAD_IDX, device)

model_conv = Seq2Seq(enc, dec).to(device)

def count_parameters(model_conv):
    return sum(p.numel() for p in model_conv.parameters() if p.requires_grad)

print(f'The model_conv has {count_parameters(model_conv):,} trainable parameters')

The model_conv has 53,420,353 trainable parameters


In [25]:
optimizer = optim.Adam(model_conv.parameters())
criterion = nn.CrossEntropyLoss(ignore_index = TRG_PAD_IDX)

In [26]:
def train(model_conv, iterator, optimizer, criterion, clip):
    
    model_conv.train()    
    epoch_loss = 0
    
    for i, batch in enumerate(iterator):
        
        src = batch.src
        trg = batch.trg        
        optimizer.zero_grad()        
        output, _ = model_conv(src, trg[:,:-1])        
        output_dim = output.shape[-1]        
        output = output.contiguous().view(-1, output_dim)
        trg = trg[:,1:].contiguous().view(-1)        
        loss = criterion(output, trg)        
        loss.backward()        
        torch.nn.utils.clip_grad_norm_(model_conv.parameters(), clip)        
        optimizer.step()        
        epoch_loss += loss.item()
        
    return epoch_loss / len(iterator)

def evaluate(model_conv, iterator, criterion):
    
    model_conv.eval()    
    epoch_loss = 0
    
    with torch.no_grad():
    
        for i, batch in enumerate(iterator):

            src = batch.src
            trg = batch.trg
            output, _ = model_conv(src, trg[:,:-1])
        
            #output = [batch size, trg len - 1, output dim]
            #trg = [batch size, trg len]

            output_dim = output.shape[-1]            
            output = output.contiguous().view(-1, output_dim)
            trg = trg[:,1:].contiguous().view(-1)           
            loss = criterion(output, trg)
            epoch_loss += loss.item()
        
    return epoch_loss / len(iterator)

In [27]:
def epoch_time(start_time, end_time):
    elapsed_time = end_time - start_time
    elapsed_mins = int(elapsed_time / 60)
    elapsed_secs = int(elapsed_time - (elapsed_mins * 60))
    return elapsed_mins, elapsed_secs

In [28]:
N_EPOCHS = 5
CLIP = 0.1

best_valid_loss = float('inf')

for epoch in range(N_EPOCHS):
    
    start_time = time.time()
    
    train_loss = train(model_conv, train_iterator, optimizer, criterion, CLIP)
    
    end_time = time.time()
    
    epoch_mins, epoch_secs = epoch_time(start_time, end_time)

    test_loss = evaluate(model_conv, test_iterator, criterion)
    torch.save(model_conv.state_dict(), 'tut5-model.pt')
    
    print(f'Epoch: {epoch+1:02} | Time: {epoch_mins}m {epoch_secs}s')
    print(f'\tTrain Loss: {train_loss:.3f} | Train PPL: {math.exp(train_loss):7.3f}')
    print(f'\t Test. Loss: {test_loss:.3f} |  Test. PPL: {math.exp(test_loss):7.3f}')

Epoch: 01 | Time: 6m 0s
	Train Loss: 3.497 | Train PPL:  33.025
	 Test. Loss: 2.302 |  Test. PPL:   9.989
Epoch: 02 | Time: 6m 0s
	Train Loss: 2.362 | Train PPL:  10.609
	 Test. Loss: 1.892 |  Test. PPL:   6.632
Epoch: 03 | Time: 6m 0s
	Train Loss: 2.050 | Train PPL:   7.765
	 Test. Loss: 1.716 |  Test. PPL:   5.561
Epoch: 04 | Time: 6m 0s
	Train Loss: 1.865 | Train PPL:   6.457
	 Test. Loss: 1.600 |  Test. PPL:   4.955
Epoch: 05 | Time: 5m 59s
	Train Loss: 1.734 | Train PPL:   5.663
	 Test. Loss: 1.524 |  Test. PPL:   4.592


# Inference: Convolutuional

# 예시 확인

In [29]:
def basic_clean_ko(text):
    text = re.split('(\d+)', text) # 숫자와 글자 분리
    return ' '.join(text)

def sent_tok_korean(text):
   # text = re.sub(r'[^\w\s]', '', text) # 구두점 제거 
    text = text.replace('  ', ' ') # 다중 공백 제거
    text = re.split('([가-힣]+[^\w\s])', text) # 구두점과 글자 분리    
    return ''.join(text) # 단어 분리   

In [30]:
## 형태소 분석을 할 경우

sen_list_morph = [
'모든 액체 , 젤 , 에어로졸 등 은 1 커트 짜리 여닫이 투명 봉지 하나 에 넣 어야 하 ㅂ니다 .',
'미안 하 지만 , 뒷쪽 아이 들 의 떠들 는 소리 가 커 어서 , 광화문 으로 가 아고 싶 은데 표 를 바꾸 어 주 시 겠 어요 ?',
'은행 이 너무 멀 어서 안 되 겠 네요 . 현찰 이 필요 하면 돈 을 훔치 시 어요',
'아무래도 분실 하 ㄴ 것 같 으니 분실 신고서 를 작성 하 아야 하 겠 습니다 . 사무실 로 같이 가 시 ㄹ 까요 ?',
'부산 에서 코로나 확진자 가 급증 하 아서 병상 이 부족하 아 지자  확진자 20명 을 대구 로 이송하 ㄴ다 .',
'변기 가 막히 었 습니다 .',
'그 바지 좀 보이 어 주 시 ㅂ시오 . 이거 얼마 에 사 ㄹ 수 있 는 것 이 ㅂ니까 ?',
'비 가 오 아서 백화점 으로 가지 말 고 두타 로 가 았 으면 좋 겠 습니다 .',
'속 이 안 좋 을 때 는 죽 이나 미음 으로 아침 을 대신 하 ㅂ니다',
'문 대통령 은 집단 이익 에서 벗어 나 아 라고 말 하 었 다 .',
'이것 좀 먹어 보 ㄹ 몇 일 간 의 시간 을 주 시 어요 .',
'이날 개미군단 은 외인 의 물량 을 모두 받 아 내 었 다 .',
'통합 우승 의 목표 를 달성하 ㄴ NC 다이노스 나성범 이 메이저리그 진출 이라는 또 다른 꿈 을 향하 어 나아가 ㄴ다 .',
'이번 구조 조정 이 제품 을 효과 적 으로 개발 하 고 판매 하 기 위하 ㄴ 회사 의 능력 강화 조처 이 ㅁ 을 이해 하 아 주 시 리라 생각 하 ㅂ니다 .',
'요즘 이 프로그램 녹화 하 며 많은 걸 느끼 ㄴ다 '
'배낭 여행 은 우리 들 이 어리었을 때 허용 되 지 않 으면 우울 하 아 하 ㄴ다'
'그 소녀 는 단지 늑대 처럼 울부짖 을 수 있 을 뿐 이 었 다']
 
sent_morph = pd.DataFrame(sen_list_morph, columns = ["sentences"])
sent_morph["clean_Korean"] = sent_morph["sentences"].apply(basic_clean_ko)
sent_morph["clean_Korean"] = sent_morph["clean_Korean"].apply(sent_tok_korean)
sent_morph = sent_morph["clean_Korean"]
sent_morph.to_csv("sent_morph")

In [31]:
reading_sent = pd.read_csv("./sent_morph")
reading_sent["trg"] = ""
reading_sent

Unnamed: 0  \
0            0   
1            1   
2            2   
3            3   
4            4   
5            5   
6            6   
7            7   
8            8   
9            9   
10          10   
11          11   
12          12   
13          13   
14          14   

                                                                                                    clean_Korean  \
0                                                       모든 액체 , 젤 , 에어로졸 등 은 1 커트 짜리 여닫이 투명 봉지 하나 에 넣 어야 하 ㅂ니다 .   
1                                      미안 하 지만 , 뒷쪽 아이 들 의 떠들 는 소리 가 커 어서 , 광화문 으로 가 아고 싶 은데 표 를 바꾸 어 주 시 겠 어요 ?   
2                                                                 은행 이 너무 멀 어서 안 되 겠 네요 . 현찰 이 필요 하면 돈 을 훔치 시 어요   
3                                              아무래도 분실 하 ㄴ 것 같 으니 분실 신고서 를 작성 하 아야 하 겠 습니다 . 사무실 로 같이 가 시 ㄹ 까요 ?   
4                                                 부산 에서 코로나 확진자 가 급증 하 아서 병상 이 부족하 아 지자 확진자 20 명 을 대구 로 이송하 ㄴ다 .   
5                                                                                                변기 가 막히 었 습니다 .   
6                                                              그 바지 좀 보이 어 주 시 ㅂ시오 . 이거 얼마 에 사 ㄹ 수 있 는 것 이 ㅂ니까 ?   
7                                                                   비 가 오 아서 백화점 으로 가지 말 고 두타 로 가 았 으면 좋 겠 습니다 .   
8                                                                         속 이 안 좋 을 때 는 죽 이나 미음 으로 아침 을 대신 하 ㅂ니다   
9                                                                           문 대통령 은 집단 이익 에서 벗어 나 아 라고 말 하 었 다 .   
10                                                                             이것 좀 먹어 보 ㄹ 몇 일 간 의 시간 을 주 시 어요 .   
11                                                                            이날 개미군단 은 외인 의 물량 을 모두 받 아 내 었 다 .   
12                                          통합 우승 의 목표 를 달성하 ㄴ NC 다이노스 나성범 이 메이저리그 진출 이라는 또 다른 꿈 을 향하 어 나아가 ㄴ다 .   
13                       이번 구조 조정 이 제품 을 효과 적 으로 개발 하 고 판매 하 기 위하 ㄴ 회사 의 능력 강화 조처 이 ㅁ 을 이해 하 아 주 시 리라 생각 하 ㅂ니다 .   
14  요즘 이 프로그램 녹화 하 며 많은 걸 느끼 ㄴ다 배낭 여행 은 우리 들 이 어리었을 때 허용 되 지 않 으면 우울 하 아 하 ㄴ다그 소녀 는 단지 늑대 처럼 울부짖 을 수 있 을 뿐 이 었 다   

   trg  
0       
1       
2       
3       
4       
5       
6       
7       
8       
9       
10      
11      
12      
13      
14

In [32]:
Raw_datafields = [("Unnamed:0", None), ("src", SRC), ("trg", TRG)]

sent_data = data.TabularDataset(
        path='./sent_morph',
        format='csv',
        skip_header=True, 
        fields=Raw_datafields) 

SRC.build_vocab(train_data, min_freq = 1)
TRG.build_vocab(train_data, min_freq = 1)

In [33]:
def translate_sentence(sentence, src_field, trg_field, model_conv, device, max_len = 50):

    model_conv.eval()
        
    if isinstance(sentence, str):
        
        tokens = [token.text.lower() for token in tok_korean(sentence)]
    else:
        tokens = [token.lower() for token in sentence]

    tokens = [src_field.init_token] + tokens + [src_field.eos_token]        
    src_indexes = [src_field.vocab.stoi[token] for token in tokens]
    src_tensor = torch.LongTensor(src_indexes).unsqueeze(0).to(device)

    with torch.no_grad():
        encoder_conved, encoder_combined = model_conv.encoder(src_tensor)

    trg_indexes = [trg_field.vocab.stoi[trg_field.init_token]]

    for i in range(max_len):

        trg_tensor = torch.LongTensor(trg_indexes).unsqueeze(0).to(device)

        with torch.no_grad():
            output, attention = model_conv.decoder(trg_tensor, encoder_conved, encoder_combined)
        
        pred_token = output.argmax(2)[:,-1].item()        
        trg_indexes.append(pred_token)

        if pred_token == trg_field.vocab.stoi[trg_field.eos_token]:
            break
    
    trg_tokens = [trg_field.vocab.itos[i] for i in trg_indexes]
    
    return trg_tokens[1:], attention

def display_attention(sentence, translation, attention):
    
    fig = plt.figure(figsize=(10,10))
    ax = fig.add_subplot(111)        
    attention = attention.squeeze(0).cpu().detach().numpy()    
    cax = ax.matshow(attention, cmap='bone')   
    ax.tick_params(labelsize=15)
    ax.set_xticklabels(['']+['<sos>']+[t.lower() for t in sentence]+['<eos>'], 
                       rotation=45)
    ax.set_yticklabels(['']+translation)
    ax.xaxis.set_major_locator(ticker.MultipleLocator(1))
    ax.yaxis.set_major_locator(ticker.MultipleLocator(1))

    plt.show()
    plt.close()

In [34]:
pred_sent = []

for i in range(len(sent_data)):
    src = vars(sent_data[i])['src']
    translation, attention = translate_sentence(src, SRC, TRG, model_conv, device)
    translation = ' '.join(translation)
    pred_sent.append(translation)
    
pred = pd.DataFrame(pred_sent, columns = ["pred_sentences"])

reading_sent["trg"] = pred["pred_sentences"]
reading_sent

Unnamed: 0  \
0            0   
1            1   
2            2   
3            3   
4            4   
5            5   
6            6   
7            7   
8            8   
9            9   
10          10   
11          11   
12          12   
13          13   
14          14   

                                                                                                    clean_Korean  \
0                                                       모든 액체 , 젤 , 에어로졸 등 은 1 커트 짜리 여닫이 투명 봉지 하나 에 넣 어야 하 ㅂ니다 .   
1                                      미안 하 지만 , 뒷쪽 아이 들 의 떠들 는 소리 가 커 어서 , 광화문 으로 가 아고 싶 은데 표 를 바꾸 어 주 시 겠 어요 ?   
2                                                                 은행 이 너무 멀 어서 안 되 겠 네요 . 현찰 이 필요 하면 돈 을 훔치 시 어요   
3                                              아무래도 분실 하 ㄴ 것 같 으니 분실 신고서 를 작성 하 아야 하 겠 습니다 . 사무실 로 같이 가 시 ㄹ 까요 ?   
4                                                 부산 에서 코로나 확진자 가 급증 하 아서 병상 이 부족하 아 지자 확진자 20 명 을 대구 로 이송하 ㄴ다 .   
5                                                                                                변기 가 막히 었 습니다 .   
6                                                              그 바지 좀 보이 어 주 시 ㅂ시오 . 이거 얼마 에 사 ㄹ 수 있 는 것 이 ㅂ니까 ?   
7                                                                   비 가 오 아서 백화점 으로 가지 말 고 두타 로 가 았 으면 좋 겠 습니다 .   
8                                                                         속 이 안 좋 을 때 는 죽 이나 미음 으로 아침 을 대신 하 ㅂ니다   
9                                                                           문 대통령 은 집단 이익 에서 벗어 나 아 라고 말 하 었 다 .   
10                                                                             이것 좀 먹어 보 ㄹ 몇 일 간 의 시간 을 주 시 어요 .   
11                                                                            이날 개미군단 은 외인 의 물량 을 모두 받 아 내 었 다 .   
12                                          통합 우승 의 목표 를 달성하 ㄴ NC 다이노스 나성범 이 메이저리그 진출 이라는 또 다른 꿈 을 향하 어 나아가 ㄴ다 .   
13                       이번 구조 조정 이 제품 을 효과 적 으로 개발 하 고 판매 하 기 위하 ㄴ 회사 의 능력 강화 조처 이 ㅁ 을 이해 하 아 주 시 리라 생각 하 ㅂ니다 .   
14  요즘 이 프로그램 녹화 하 며 많은 걸 느끼 ㄴ다 배낭 여행 은 우리 들 이 어리었을 때 허용 되 지 않 으면 우울 하 아 하 ㄴ다그 소녀 는 단지 늑대 처럼 울부짖 을 수 있 을 뿐 이 었 다   

                                                                                                                                    trg  
0   all liquids , gels , gels , aerosols , aerosols , aerosols , aerosols , aerosols , aerosols , aerosols , zip-top , aerosols . <eos>  
1                               i am sorry , but i want to change the noise , would you please change the ticket to gwanghwamun ? <eos>  
2                                                                               the bank is too far . you can have a cash money . <eos>  
3                                               i think you have lost the lost report , i will have to fill out the office form ? <eos>  
4                                           the food is the most specialty in busan , and the keyboard is the best of the busan . <eos>  
5                                                                                 the toilet is clogged . the toilet is clogged . <eos>  
6                                                                 please show me the pants , how much is this to buy this pants ? <eos>  
7                                                                    i hope it will rain to the department store and go to duta . <eos>  
8                                                  i am afraid i ca not make a little in the morning , i will be in the morning . <eos>  
9                                                                     the doors told me the president of the national president . <eos>  
10                                  i would like to have a few time to get this of the time to get this of the time to see this . <eos>  
11                                                                                       the shipping was the loss of oeinmyoji . <eos>  
12                                     the baggage of the main ,

In [35]:
## 어절단위로 결합하여 할 경우

sen_list_eojeol = [
'모든 액체, 젤, 에어로졸 등은 1커트 짜리 여닫이 투명 봉지 하나에 넣어야 합니다.',
'미안하지만 , 뒷쪽 아이들의 떠드는 소리가 커서 , 광화문으로 가고 싶은데 표를 바꾸어 주시겠어요?',
'은행이 너무 멀어서 안되겠네요. 현찰이 필요하면 돈을 훔치세요',
'아무래도 분실한 것 같으니 분실신고서를 작성해야 하겠습니다. 사무실로 같이 가실까요?',
'부산에서 코로나 확진자가 급증해서 병상이 부족해지자  확진자 20명을 대구로 이송한다.',
'변기가 막히었습니다 .',
'그 바지 좀 보여주십시오. 이거 얼마에 살 수 있는 것입니까?',
'비가 와서 백화점으로 가지 말고 두타로 갔으면 좋겠습니다 .',
'속이 안좋을 때는 죽이나 미음으로 아침을 대신 합니다',
'문 대통령은 집단이익에서 벗어나라고 말하였다.',
'이것 좀 먹어 볼 몇 일 간의 시간을 주세요.',
'이날 개미군단은 외인의 물량을 모두 받아내었다.',
'통합 우승의 목표를 달성한 NC 다이노스 나성범이 메이저리그 진출이라는 또다른 꿈을 향해 나아간다.',
'이번 구조 조정이 제품을 효과적으로 개발하고 판매하기 위한 회사의 능력 강화 조처임을 이해해 주시리라 생각합니다.',
'요즘 이 프로그램 녹화하며 많은 걸 느낀다'
'배낭 여행은 우리들이 어렸을 때 허용디지 않으면 우울 해 한다'
'그 소녀는 단지 늑대처럼 울부짖을 수 있을 뿐이었다']
sent_eojeol = pd.DataFrame(sen_list_eojeol, columns = ["sentences"])
sent_eojeol["clean_Korean"] = sent_eojeol["sentences"].apply(basic_clean_ko)
sent_eojeol["clean_Korean"] = sent_eojeol["clean_Korean"].apply(sent_tok_korean)
sent_eojeol = sent_eojeol["clean_Korean"]
sent_eojeol.to_csv("sent_eojeol")

In [36]:
reading_sent2 = pd.read_csv("./sent_eojeol")
reading_sent2["trg"] = ""

Raw_datafields = [("Unnamed:0", None), ("src", SRC), ("trg", TRG)]

sent_data2 = data.TabularDataset(
        path='./sent_eojeol',
        format='csv',
        skip_header=True, 
        fields=Raw_datafields) 

SRC.build_vocab(train_data, min_freq = 1)
TRG.build_vocab(train_data, min_freq = 1)

In [37]:
pred_sent = []

for i in range(len(sent_data2)):
    src = vars(sent_data2[i])['src']
    translation, attention = translate_sentence(src, SRC, TRG, model_conv, device)
    translation = ' '.join(translation)
    pred_sent.append(translation)
    
pred = pd.DataFrame(pred_sent, columns = ["pred_sentences"])

reading_sent2["trg"] = pred["pred_sentences"]
reading_sent2

Unnamed: 0  \
0            0   
1            1   
2            2   
3            3   
4            4   
5            5   
6            6   
7            7   
8            8   
9            9   
10          10   
11          11   
12          12   
13          13   
14          14   

                                                                             clean_Korean  \
0                                        모든 액체, 젤, 에어로졸 등은 1 커트 짜리 여닫이 투명 봉지 하나에 넣어야 합니다.   
1                                 미안하지만 , 뒷쪽 아이들의 떠드는 소리가 커서 , 광화문으로 가고 싶은데 표를 바꾸어 주시겠어요?   
2                                                      은행이 너무 멀어서 안되겠네요. 현찰이 필요하면 돈을 훔치세요   
3                                         아무래도 분실한 것 같으니 분실신고서를 작성해야 하겠습니다. 사무실로 같이 가실까요?   
4                                        부산에서 코로나 확진자가 급증해서 병상이 부족해지자 확진자 20 명을 대구로 이송한다.   
5                                                                            변기가 막히었습니다 .   
6                                                      그 바지 좀 보여주십시오. 이거 얼마에 살 수 있는 것입니까?   
7                                                       비가 와서 백화점으로 가지 말고 두타로 갔으면 좋겠습니다 .   
8                                                           속이 안좋을 때는 죽이나 미음으로 아침을 대신 합니다   
9                                                               문 대통령은 집단이익에서 벗어나라고 말하였다.   
10                                                              이것 좀 먹어 볼 몇 일 간의 시간을 주세요.   
11                                                             이날 개미군단은 외인의 물량을 모두 받아내었다.   
12                                통합 우승의 목표를 달성한 NC 다이노스 나성범이 메이저리그 진출이라는 또다른 꿈을 향해 나아간다.   
13                        이번 구조 조정이 제품을 효과적으로 개발하고 판매하기 위한 회사의 능력 강화 조처임을 이해해 주시리라 생각합니다.   
14  요즘 이 프로그램 녹화하며 많은 걸 느낀다배낭 여행은 우리들이 어렸을 때 허용디지 않으면 우울 해 한다그 소녀는 단지 늑대처럼 울부짖을 수 있을 뿐이었다   

                                                                                    trg  
0   all the specialty are all made for the main 1 and a pair of food , the food ? <eos>  
1                                          do you have any special or specialty ? <eos>  
2                                              is this specialty made specialty ? <eos>  
3                                    do you have any special food with the food ? <eos>  
4                           is this special special special food food specialty ? <eos>  
5                                                         this food are working . <eos>  
6                                                      can i buy a pair of food ? <eos>  
7                                       the food food you have the special food ? <eos>  
8                                             do you have any food or specialty ? <eos>  
9                                          do you have the baggage baggage food ? <eos>  
10                                    how many things are this in the main area ? <eos>  
11                                             is the specialty all the baggage ? <eos>  
12                    is it made of special specialty specialty in your baggage ? <eos>  
13                                   is this the baggage food for the specialty ? <eos>  
14             is this made working working with the specialty of the main food ? <eos>

In [38]:
from torchtext.data.metrics import bleu_score

def calculate_bleu(data, src_field, trg_field, model_conv, device, max_len = 50):
    
    trgs = []
    pred_trgs = []
    
    for datum in data:
        
        src = vars(datum)['src']
        trg = vars(datum)['trg']
        
        pred_trg, _ = translate_sentence(src, src_field, trg_field, model_conv, device, max_len)
        
        #cut off <eos> token
        pred_trg = pred_trg[:-1]
        
        pred_trgs.append(pred_trg)
        trgs.append([trg])
        
    return bleu_score(pred_trgs, trgs)

In [39]:
bleu_score = calculate_bleu(test_data, SRC, TRG,model_conv, device)

print(f'BLEU score = {bleu_score*100:.2f}')

BLEU score = 27.35


# 정리

BATCH_SIZE = 256

N-EPOCH = 5

Train_PPL = 5.663

Test_PPL = 4.592

BLUE score = 27.36